design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,310 @@
+return a SUT that can be used to authenticate the user to Conjur.
+
+The format of the URL matches the existing authenticator endpoint for
+authn-ldap.
+
+##### `POST /authn-sut/<account>/authenticate`
+
+This endpoint will be used to authenticate a user to Conjur using a SUT. The
+user will provide the SUT in the request body. The endpoint will return a
+short-lived auth token.
+
+The format of the URL matches the existing authenticator endpoint for
+authn-ldap.
+
+### Low Level Design
+
+#### Data Model
+
+The SUT will be a cryptographically secure, random string. It will be generated
+by the Conjur server and stored in the database. The SUT will be associated with
+the user and expiration time. The SUT will be marked as used once it is utilized
+to authenticate the user to Conjur.
+
+When we store the SUT in the database, we will store a hash of the SUT instead
+of the SUT itself. This will prevent an attacker with access to the database
+from being able to use the SUTs to authenticate to Conjur. The SUT will be
+hashed using a cryptographically secure hash function, such as SHA-256.
+When a user presents a SUT to authenticate to Conjur, we will hash the SUT and
+compare it to the hash stored in the database.
+
+##### SUT Format
+
+We cannot use JWTs for the SUT as we need the ability to revoke the SUT after it
+is used, a feature not easily implemented with JWTs. This would make it
+impossible to prevent replay attacks.
+
+##### Open Questions
+
+- Is it possible to restrict use of the SUT to a specific IP range (like with
+  hostfactory tokens) or other criteria to limit risk if the SUT is leaked?
+- Is it better to actually delete the SUT from the database once it is used or
+  to just mark it as used?
+- Is it better to store a `used_at` timestamp or a boolean `used` flag?
+- Should the SUT expiration be configurable (via policy) or should it be a fixed
+  value? What should the default expiration be? (30 seconds?)
+
+##### Example Database Table
+
+| id | role_id | guid_hash | expires_at | used_at |
+|----|---------|-----------|------------|---------|
+| 1  | user:john.doe | lkduj...  | 2023-09-28 12:08:23 | NULL |
+| 2  | user:demo.user | 2k1g2...  | 2023-09-28 12:09:45 | 2023-09-28 12:09:10 |
+
+For this example, we would need a migration to add the `authn_sut_tokens` table.
+That would look something like this:
+
+```ruby
+Sequel.migration do
+  change do
+    create_table :authn_sut_tokens do
+      primary_key :id
+      foreign_key :role_id, :roles, type: String, null: false, on_delete: :cascade
+      String :guid_hash, null: false, unique: true
+      DateTime :expires_at, null: false
+      DateTime :used_at
+    end
+  end
+end
+```
+
+#### Authenticator
+
+The SUT authenticator will be implemented as a new authenticator plugin. The
+implementation will follow the existing authenticator plugin pattern, as
+documented in [AUTHENTICATORS.md](../AUTHENTICATORS.md)
+
+## Security
```

Comment:
I'm thinking about taking a page from OIDC to add security here. What if PSM (or any client) had to generate a unique code and passed the hash of that code to the authn-sut/login endpoint. Then, PSM sends the actual code to Conjur UI. Conjur UI includes that along with the SUT to Conjur. Conjur applies the same hash (we'd have to pick a single algorithm - maybe SHA256) to the code and only honors the SUT if that matches the request code.

Reply:
Filling this out more explicitly - I'm modeling this on PKCE in OIDC. 
1) PSM generates a GUID for a specific single use token request.
2) PSM SHA256 hashes the GUID and sends the resulting hash value and the transformation method (currently only "sha256" supported, but this will make it easier for us to introduce post-quantum crypto handling or other algorithmss later) to the endpoint for the SUT.
3) The single use endpoint includes the received hash value and transformation methods as additional columns in the database where the token is stored. The token is returned to PSM.
4) PSM sends the token and the original (unhashed) GUID to Conjur UI / Conjur. The verification process includes a step that takes the GUID, applies the transformation method associated with the token, and verifies that it gets the same result as was written from the original request. If it does not get the same result, the token is rejected and declared invalid immediately. 

Key pieces:
- The GUID itself is never sent anywhere until it's used with the token to prevent interception.
- The GUID ties a single use token to a particular request, so the token can't be used elsewhere.
- The transformation method must be one way, so even if someone got a hold of the database, they couldn't reverse engineer the GUID and gain access to using the token. 
- Even though we only support one transformation method at the start, we have the mechanism in place to allow other choices in the future to make upgrades and introducting new functionality easier.